core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/source/apis/twitter/v2/TwitterFullArchiveSearchSourceOpExec.scala

```diff
@@ -0,0 +1,116 @@
+
+  // nextToken is used to retrieve next page of results, if exists.
+  var nextToken: String = _
+
+  // contains tweets from the previous request.
+  var tweetCache: mutable.Buffer[Tweet] = mutable.Buffer()
+
+  var hasNextRequest: Boolean = curLimit > 0
+
+  override def produceTexeraTuple(): Iterator[Tuple] =
+    new Iterator[Tuple]() {
+      override def hasNext: Boolean = (hasNextRequest || tweetCache.nonEmpty) && curLimit > 0
+
+      override def next: Tuple = {
+        // if the current cache is exhausted, query for the next response
+        if (tweetCache.isEmpty && hasNextRequest) {
+          queryForNextBatch(
+            searchQuery,
+            LocalDateTime.parse(fromDateTime, DateTimeFormatter.ISO_DATE_TIME),
+            LocalDateTime.parse(toDateTime, DateTimeFormatter.ISO_DATE_TIME),
+            curLimit.min(TWITTER_API_BATCH_SIZE)
+          )
+        }
+
+        // if the request is emtpy, it indicates no more tweets, iterator should stop
+        if (tweetCache.isEmpty) {
+          return null
+        }
+        val tweet: Tweet = tweetCache.remove(0)
+
+        curLimit -= 1
+
+        // if limit is 0, then no more requests should be sent
+        if (curLimit == 0) {
+          hasNextRequest = false
+        }
+
+        val fields = AttributeTypeUtils.parseFields(
+          Array[Object](
+            tweet.getId,
+            tweet.getText,
+            // given the fact that the redouane59/twittered library is using LocalDateTime as the API parameter,
+            // we have to fix it to UTC time zone to normalize the time.
+            tweet.getCreatedAt
+              .atZone(ZoneId.systemDefault())
+              .withZoneSameInstant(ZoneId.of("UTC"))
+              .toLocalDateTime
+              .atOffset(ZoneOffset.UTC)
+              .format(DateTimeFormatter.ISO_DATE_TIME),
+            tweet.getAuthorId,
+            // tweet.getUser, // currently unsupported by the redouane59/twittered library
+            // TODO: add user information
+            tweet.getLang,
+            tweet.getTweetType.toString,
+            // TODO: add actual geo related information
+            Option(tweet.getGeo).map(_.getPlaceId).orNull,
+            Option(tweet.getGeo).map(_.getCoordinates).orNull,
+            tweet.getInReplyToStatusId,
+            tweet.getInReplyToUserId,
+            Integer.valueOf(tweet.getLikeCount),
+            Integer.valueOf(tweet.getQuoteCount),
+            Integer.valueOf(tweet.getReplyCount),
+            Integer.valueOf(tweet.getRetweetCount)
+          ),
+          schema.getAttributes.map((attribute: Attribute) => { attribute.getType }).toArray
+        )
+        Tuple.newBuilder.add(schema, fields).build
+      }
+    }
+
+  private def queryForNextBatch(
+      query: String,
+      startDateTime: LocalDateTime,
+      endDateTime: LocalDateTime,
+      maxResults: Int
+  ): Unit = {
+
+    val response = twitterClient.searchForTweetsFullArchive(
```

Comment:
are there any timeout mechanism?

Reply:
we rely on the client to handle timeout, or any other error cases.